size audit workflow

amartya-dev opened the pull request:
## Proposed changes

This PR creates the plugin size audit workflow by integrating proven GitHub Actions marketplace tools. The workflow uses `preactjs/compressed-size-action` for bundle analysis, `calibreapp/image-actions` for image optimization and reporting, and `actions/size-limit` for file size monitoring. This approach reduces maintenance overhead while providing more robust and feature-rich size auditing capabilities.

## Type of Change

#### Production

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Dependency update
- [x] Refactoring / housekeeping (changes to files not directly related to functionality)

#### Development

- [ ] Tests
- [x] Dependency update
- [x] Environment update / refactoring
- [ ] Documentation Update

## Video

<!-- Add a short video demonstrating where the change takes effect and how to can be reproduced by reviewers -->
<!-- On macOS press cmd+shift+5 to open the screen recording tool. It will save the video to the desktop  -->

<!-- Drag and drop the video file into the GitHub editor to attach it -->

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/bluehost/.github/blob/master/.github/contributing.md) doc
- [x] I have viewed my change in a web-browser
- [x] Linting and tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

## Pre-deployment Checklist

- [ ] Verify that the marketplace actions (`preactjs/compressed-size-action@v2`, `calibreapp/image-actions@main`, `actions/size-limit@v1`) are compatible with our current GitHub Actions setup
- [ ] Ensure the `NEWFOLD_ACCESS_TOKEN` secret is properly configured for npm registry access
- [ ] Test the workflow on a feature branch to ensure all marketplace actions function correctly

## Post-deployment Checklist

How can the change be verified?

- [ ] Create a test PR to verify the enhanced size audit workflow runs successfully
- [ ] Check that bundle size analysis comments are posted to PRs with build files
- [ ] Verify image optimization suggestions are provided for large images
- [ ] Confirm file size monitoring reports are generated correctly
- [ ] Validate that the custom plugin size calculation and comparison logic still works as expected

circlecube commented:
@amartya-dev This looks like a good start.

Couple things:
- I am not seeing any report on the image optimization step, is that part working?
- I think it'd be best to move this workflow to the workflows repo and call it in a workflow here, then it can be shared across all plugins. Happy to help get that set up if you need.

desrosj reviewed (changes requested):
👋 I left some initial feedback!

amartya-dev commented on .github/workflows/plugin-size-audit.yml:
```diff
@@ -0,0 +1,410 @@
+
+          echo "" >> final-report.md
+
+          echo "### Largest PHP Files" >> final-report.md
```
Updated the conditional logic to check if relevant file types (.php, .js, .css, .json, .lock) are actually being modified before showing the largest files sections. Now it only shows this information when it's actually relevant to the changes being made, reducing noise in PR comments."

amartya-dev commented on .github/workflows/plugin-size-audit.yml:
```diff
@@ -0,0 +1,410 @@
+          # Store for comparison
+
+          echo $TOTAL_PHP_SIZE_BYTES > current_php_size.txt
+      - name: Calculate JavaScript Size
```
removed the redundant JavaScript/Asset Analysis section and related size calculation steps.

amartya-dev commented on .github/workflows/plugin-size-audit.yml:
```diff
@@ -0,0 +1,410 @@
+          repo-token: '${{ secrets.GITHUB_TOKEN }}'
+          pattern: 'build/**/*.{js,css}'
+          compression: gzip
+      - name: Image Analysis
```
Tried to add a separate workflow to make another optimisation pr

amartya-dev commented on .github/workflows/plugin-size-audit.yml:
```diff
@@ -0,0 +1,410 @@
+
+          # Store for comparison
+
+          echo $TOTAL_SIZE_BYTES > current_size.txt
```
Replaced all temporary text files with proper step outputs using $GITHUB_OUTPUT. Now values are referenced using steps.plugin_size.outputs.total_size_bytes etc. instead of reading from text files

amartya-dev commented on .github/workflows/plugin-size-audit.yml:
```diff
@@ -0,0 +1,410 @@
+          compressOnly: true
+          jpegQuality: '60'
+          pngQuality: '60'
+          webpQuality: '60'
```
Updated image quality settings to match WordPress Core standards: JPEG/PNG quality from 60 to 82, WebP quality from 60 to 86.
You're absolutely right to question these values. This was work in progress - I initially set them at 60 just to test how the image optimization was working, but had to move to a different task before properly researching the optimal values.
The WordPress Core team's values of 82/86 make much more sense and I should have done this research upfront instead of using placeholder values. Thanks for catching this and pointing me to the established standards.

amartya-dev commented on .github/workflows/plugin-size-audit.yml:
```diff
@@ -0,0 +1,410 @@
+name: Plugin Size and Asset Audit
+'on':
```
✅ Implemented - Fixed YAML formatting by removing unnecessary quotes around on: and adding quotes around branch names for consistency.

amartya-dev commented on .github/workflows/plugin-size-audit.yml:
```diff
@@ -0,0 +1,410 @@
+              owner: context.repo.owner,
+              repo: context.repo.repo,
+              body: report
+            });
```
✅ Implemented - Added logic to update existing comments instead of creating new ones on each run. This prevents comment spam by identifying and updating the previous workflow comment.

Note on fork limitation: You're correct that this won't work for forked repositories due to GitHub's security restrictions. Since this is a public repo, we should consider this limitation. Would you like me to add a condition to skip commenting on forked PRs, or do we want to keep it as-is given the low likelihood of external contributors?"

amartya-dev commented on .github/workflows/plugin-size-audit.yml:
```diff
@@ -0,0 +1,410 @@
+      - name: Install NPM Dependencies
+        run: npm install --legacy-peer-deps
+      - name: Build JavaScript
+        run: npm run build
```
✅ Implemented - Removed the unnecessary 'Build JavaScript' step. The preactjs/compressed-size-action handles building automatically, as confirmed by the WordPress Gutenberg repository usage.
Why I initially included it: I was being overly cautious and following a traditional explicit CI/CD pattern, but should have trusted the action's documentation and established patterns from the WordPress Gutenberg repo."

amartya-dev commented on .github/workflows/plugin-size-audit.yml:
```diff
@@ -0,0 +1,410 @@
+          name: size-audit-reports
+          path: final-report.md
+          retention-days: 90
+      - name: Fail on Size Violations
```
I completely removed the "Fail on Size Violations" step that would explicitly fail the workflow

amartya-dev commented on .github/workflows/plugin-size-audit.yml:
```diff
@@ -0,0 +1,410 @@
+      - name: Setup Node.js
+        uses: actions/setup-node@v4
+        with:
+          node-version: 20.x
```
I replaced the hardcoded Node.js version with the node-version-file input that references the .nvmrc file

amartya-dev commented on .github/workflows/plugin-size-audit.yml:
```diff
@@ -0,0 +1,410 @@
+        uses: actions/checkout@v4
+        with:
+          fetch-depth: 0
+      - name: Setup PHP
```
done

amartya-dev commented on .github/workflows/plugin-size-audit.yml:
```diff
@@ -0,0 +1,410 @@
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v4
```
done

desrosj reviewed (changes requested):
Left a new round of feedback.

It's not currently set up in this repository, but I've [run Zizmor locally](https://docs.zizmor.sh/) and there are several findings to address. I've noted some of them contextually within the files.

One that needs to be addressed in several locations is related to potential code injection attacks. Zizmor has a [good documentation page](https://docs.zizmor.sh/audits/#template-injection) detailing the risks, and GitHub also has a [detailed blog post about the problem with remediation suggestions](https://securitylab.github.com/resources/github-actions-untrusted-input/).

The TL;DR is that any time a value stored in a `steps.*.outputs.*` output variable is used, it [should be passed to the step as an `env` variable instead](https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_idstepsenv) and referenced using `${VARNAME}` instead (not `${{ env.VARNAME }}`. There were 11 instances of this flagged in this PR by Zizmor.

desrosj commented:
@amartya-dev I've merged #176 and synced the changes into this PR. It's now pointing out a lot of the issues that I mentioned automatically. That should help refined the code to be more secure. Let me know if you have any questions!

amartya-dev commented:
- Pinned all third-party actions to commit SHAs and updated to safe SHAs.
- Removed unsafe use of steps outputs; pass via env and reference with ${VAR}.
- Built assets once; compared against base via a detached worktree (no custom-ref checkout).
- Switched to npm ci; sanitized and removed credentials (.npmrc and git extraheader).
- Included vendor in total plugin size; excluded node_modules; fixed size math (bytes → human).
- Added thresholds to fail on size regressions; added image-size check.
- Consolidated reporting: single PR comment + artifact; added JS/CSS build size and deltas.
- Hardened checkouts (fetch-depth: 1, persist-credentials: false); added least-privilege permissions.
- Extended triggers (PRs to main and develop); fixed ShellCheck lint across scripts.
- Image optimization workflow: scoped permissions, hardened checkout, sanitized creds, removed unnecessary token input.
